source/slang/slang-ast-serialize.cpp

```diff
@@ -46,6 +49,60 @@ static void _toNativeValue(ASTSerialReader* reader, const SERIAL_TYPE& src, NATI
     ASTSerialTypeInfo<NATIVE_TYPE>::toNative(reader, &src, &dst);
 }
 
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! ModuleASTSerialFilter  !!!!!!!!!!!!!!!!!!!!!!!!
+
+
+ASTSerialIndex ModuleASTSerialFilter::writePointer(ASTSerialWriter* writer, const NodeBase* inPtr) 
+{
+    NodeBase* ptr = const_cast<NodeBase*>(inPtr);
+    SLANG_ASSERT(ptr);
+    
+    if (Decl* decl = as<Decl>(ptr))
+    {
+        ModuleDecl* moduleDecl = findModuleForDecl(decl);
+        SLANG_ASSERT(moduleDecl);
+        if (moduleDecl && moduleDecl != m_moduleDecl)
+        {
+            ASTBuilder* astBuilder = m_moduleDecl->module->getASTBuilder();
+
+            // It's a reference to a declaration in another module, so create an ImportExternalDecl.
+
+            String mangledName = getMangledName(astBuilder, decl);
+
+            ImportExternalDecl* importDecl = astBuilder->create<ImportExternalDecl>();
+            importDecl->mangledName = mangledName;
+            const ASTSerialIndex index = writer->writePointer(importDecl);
+
+            // Set as the index of this
+            writer->setPointerIndex(ptr, index);
+
+            return index;
+        }
+        else
+        {
+            // Okay... we can just write it out then
+            return writer->writePointer(ptr);
+        }
+    }
+
+    // TODO(JS): What we really want to do here is to ignore bodies functions.
+    // It's not 100% clear if this is even right though - for example does type inference
+    // imply the body is needed to say infer a return type?
+    // Also not clear if statements in other scenarios (if there are others) might need to be kept.
+    //
+    // For now we just ignore all stmts
```

Comment:
Given our current setup, I believe it is valid to skip all `Stmt`s when serializing, because they only affect the "body" of various declarations.

Right now we don't infer things like the return type of a function from its body, but even if we did I would expect AST serialization to be done *after* the inference has run and the result type has been recorded on the function declaration itself.

The main sticking point I anticipate for serialization right now comes from `Expr`s instead of `Stmt`s.. An expression can end up in the AST in a context that isn't just a function "body" in two ways:

1. A type can in principle include arbitrary expression as a generic argument (`vector<float, 1 + foo(N)>`). Right now those are basked down to `Val`s, and we don't support `Val`s that are full expressions, so this is a non-issue today just because of other limitations in the compiler.

2. A function can include a default-value expression for one or more of its parameters, and these are represented as a bare `Expr` today.

Item (2) is the main reason that I expect we need to have serialization of `Expr`s in the near term, because lowering a *call* to a function in another module currently requires being able to emit the value of its default arguments as needed. Something more clever is possible, but we'd have to change how we do code generation for default-value expressions.